- The reason for adding `tests/lib/RealRepairTest.php` is tests/lib/RepairStepTest.php because `RepairStepTest.php` fails by default. For example if we run:

```
➜  owncloud3 git:(master) ✗ make test-php TEST_DATABASE=sqlite TEST_PHP_SUITE=tests/lib/RepairStepTest.php                     
php build/composer.phar install --dev
Cannot load Xdebug - it was already loaded
You are using the deprecated option "dev". Dev packages are installed by default now.
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Nothing to install or update
Package guzzle/common is abandoned, you should avoid using it. Use guzzle/guzzle instead.
Package guzzle/http is abandoned, you should avoid using it. Use guzzle/guzzle instead.
Package guzzle/parser is abandoned, you should avoid using it. Use guzzle/guzzle instead.
Package guzzle/stream is abandoned, you should avoid using it. Use guzzle/guzzle instead.
Generating optimized autoload files
ocramius/package-versions:  Generating version class...
ocramius/package-versions: ...done generating version class
PHPUNIT="/home/sujith/test/owncloud3/lib/composer/phpunit/phpunit/phpunit" build/autotest.sh sqlite tests/lib/RepairStepTest.php
Using PHP executable /usr/bin/php
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Parsing all files in lib/public for the presence of @since or @deprecated on each method...

Using database oc_autotest
Setup environment for sqlite testing on local storage ...
Installing ....
Cannot load Xdebug - it was already loaded
creating sqlite db
ownCloud was successfully installed
Testing with sqlite ...
Cannot load Xdebug - it was already loaded
/home/sujith/test/owncloud3/lib/composer/phpunit/phpunit/phpunit --configuration phpunit-autotest.xml --coverage-clover autotest-clover-sqlite.xml --coverage-html coverage-html-sqlite --log-junit autotest-results-sqlite.xml ../tests/lib/RepairStepTest.php 
Cannot load Xdebug - it was already loaded
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.6-1+ubuntu16.04.1+deb.sury.org+1 with Xdebug 2.6.0
Configuration: /home/sujith/test/owncloud3/tests/phpunit-autotest.xml

EEWE                                                                4 / 4 (100%)

Time: 5.93 seconds, Memory: 38.00MB

There were 3 errors:

1) Test\RepairTest::testRunRepairStep
Error: Class 'Test\TestRepairStep' not found

/home/sujith/test/owncloud3/tests/lib/RepairStepTest.php:61

2) Test\RepairTest::testRunRepairStepThatFail
Error: Class 'Test\TestRepairStep' not found

/home/sujith/test/owncloud3/tests/lib/RepairStepTest.php:74

3) Test\RepairTest::testRunRepairStepsContinueAfterWarning
Error: Class 'Test\TestRepairStep' not found

/home/sujith/test/owncloud3/tests/lib/RepairStepTest.php:116

--

There was 1 warning:

1) Test\RepairTest::testRunRepairStepsWithException
Cannot stub or mock class or interface "\Test\TestRepairStep" which does not exist

ERRORS!
Tests: 4, Assertions: 0, Errors: 3, Warnings: 1.

Generating code coverage report in Clover XML format ...^CMakefile:163: recipe for target 'test-php' failed
make: *** [test-php] Interrupt

➜  owncloud3 git:(master) ✗
```
That's the reason I quickly wrote `tests/lib/RealRepairTest.php`.

--- a/apps/files/lib/Command/Scan.php
+++ b/apps/files/lib/Command/Scan.php
@@ -160,11 +160,12 @@ class Scan extends Base {
 		$repairStep = new RepairMismatchFileCachePath(
 			$connection,
 			$this->mimeTypeLoader,
-			$this->logger
+			$this->logger,
+			$this->config
 		);
 		$repairStep->setStorageNumericId(null);
 		$repairStep->setCountOnly(false);
-		$repairStep->run(new ConsoleOutput($output));
+		$repairStep->doRepair(new ConsoleOutput($output));
 	}
 
 	protected function scanFiles($user, $path, $verbose, OutputInterface $output, $backgroundScan = false, $shouldRepair = false) {
@@ -183,7 +184,8 @@ class Scan extends Base {
 					$repairStep = new RepairMismatchFileCachePath(
 						$connection,
 						$this->mimeTypeLoader,
-						$this->logger
+						$this->logger,
+						$this->config
 					);
 					$repairStep->setStorageNumericId($storage->getCache()->getNumericStorageId());
 					$repairStep->setCountOnly(false);

--- a/lib/private/Repair.php
+++ b/lib/private/Repair.php
@@ -132,7 +132,8 @@ class Repair implements IOutput {
 			new RepairMismatchFileCachePath(
 				\OC::$server->getDatabaseConnection(),
 				\OC::$server->getMimeTypeLoader(),
-				\OC::$server->getLogger()),
+				\OC::$server->getLogger(),
+				\OC::$server->getConfig()),
 			new FillETags(\OC::$server->getDatabaseConnection()),
 			new CleanTags(\OC::$server->getDatabaseConnection(), \OC::$server->getUserManager()),
 			new DropOldTables(\OC::$server->getDatabaseConnection()),

--- a/lib/private/Repair/RepairMismatchFileCachePath.php
+++ b/lib/private/Repair/RepairMismatchFileCachePath.php
@@ -21,6 +21,7 @@
 
 namespace OC\Repair;
 
+use OCP\IConfig;
 use OCP\ILogger;
 use OCP\Migration\IOutput;
 use OCP\Migration\IRepairStep;
@@ -57,15 +58,20 @@ class RepairMismatchFileCachePath implements IRepairStep {
 	/** @var ILogger  */
 	protected $logger;
 
+	/** @var IConfig */
+	protected $config;
+
 	/**
 	 * @param \OCP\IDBConnection $connection
 	 */
 	public function __construct(IDBConnection $connection,
 								IMimeTypeLoader $mimeLoader,
-								ILogger $logger) {
+								ILogger $logger,
+								IConfig $config) {
 		$this->connection = $connection;
 		$this->mimeLoader = $mimeLoader;
 		$this->logger = $logger;
+		$this->config = $config;
 	}
 
 	public function getName() {
@@ -532,11 +538,13 @@ class RepairMismatchFileCachePath implements IRepairStep {
 	}
 
 	/**
-	 * Run the repair step
+	 * The purpose of this function is to let execute the run method
+	 * irrespective of version. For example when triggered from files:scan
+	 * this repair step shouldn't be blocked.
 	 *
-	 * @param IOutput $out output
+	 * @param IOutput $out
 	 */
-	public function run(IOutput $out) {
+	public function doRepair(IOutput $out) {
 		$this->dirMimeTypeId = $this->mimeLoader->getId('httpd/unix-directory');
 		$this->dirMimePartId = $this->mimeLoader->getId('httpd');
 
@@ -568,6 +576,21 @@ class RepairMismatchFileCachePath implements IRepairStep {
 			}
 			$out->finishProgress();
 			$out->info('');
+		}
+	}
+
+	/**
+	 * Run the repair step
+	 *
+	 * @param IOutput $out output
+	 */
+	public function run(IOutput $out) {
+		$currentVersion = $this->config->getSystemValue('version', '0.0.0');
+		$versionCompareStatus = \version_compare($currentVersion, '10.0.4', '<');
+		//Execute repair step if version is less than 10.0.4 during upgrade
+		//This is not applicable when called from file scan command
+		if ($versionCompareStatus) {
+			$this->doRepair($out);
 		}
 	}
 }

--- a/tests/lib/Repair/RepairMismatchFileCachePathTest.php
+++ b/tests/lib/Repair/RepairMismatchFileCachePathTest.php
@@ -10,6 +10,7 @@ namespace Test\Repair;
 
 use OC\Repair\RepairMismatchFileCachePath;
 use OCP\Files\IMimeTypeLoader;
+use OCP\IConfig;
 use OCP\Migration\IOutput;
 use OCP\Migration\IRepairStep;
 use Test\TestCase;
@@ -30,6 +31,7 @@ class RepairMismatchFileCachePathTest extends TestCase {
 	/** @var \OCP\IDBConnection */
 	private $connection;
 
+	private $config;
 	protected function setUp() {
 		parent::setUp();
 
@@ -44,7 +46,8 @@ class RepairMismatchFileCachePathTest extends TestCase {
 
 		/** @var \PHPUnit_Framework_MockObject_MockObject | ILogger $logger */
 		$logger = $this->createMock(ILogger::class);
-		$this->repair = new RepairMismatchFileCachePath($this->connection, $mimeLoader, $logger);
+		$this->config = $this->createMock(IConfig::class);
+		$this->repair = new RepairMismatchFileCachePath($this->connection, $mimeLoader, $logger, $this->config);
 		$this->repair->setCountOnly(false);
 	}
 
@@ -209,6 +212,10 @@ class RepairMismatchFileCachePathTest extends TestCase {
 		$doNotTouchId = $this->createFileCacheEntry($sourceStorageId, 'files/source/do_not_touch', $sourceId);
 
 		$outputMock = $this->createMock(IOutput::class);
+		$this->config->expects($this->any())
+			->method('getSystemValue')
+			->with('version', '0.0.0')
+			->willReturn('10.0.3');
 		if ($repairStoragesOrder === null) {
 			// no storage selected, full repair
 			$this->repair->setStorageNumericId(null);
@@ -348,6 +355,10 @@ class RepairMismatchFileCachePathTest extends TestCase {
 
 		$outputMock = $this->createMock(IOutput::class);
 		$this->repair->setStorageNumericId($storageId);
+		$this->config->expects($this->any())
+			->method('getSystemValue')
+			->with('version', '0.0.0')
+			->willReturn('10.0.3');
 		$this->repair->run($outputMock);
 
 		// self-referencing updated
@@ -537,6 +548,10 @@ class RepairMismatchFileCachePathTest extends TestCase {
 
 		$outputMock = $this->createMock(IOutput::class);
 		$this->repair->setStorageNumericId($storageId);
+		$this->config->expects($this->any())
+			->method('getSystemValue')
+			->with('version', '0.0.0')
+			->willReturn('10.0.3');
 		$this->repair->run($outputMock);
 
 		// wrong parent root reparented to actual root
@@ -595,6 +610,10 @@ class RepairMismatchFileCachePathTest extends TestCase {
 
 		$outputMock = $this->createMock(IOutput::class);
 		$this->repair->setStorageNumericId($storageId);
+		$this->config->expects($this->any())
+			->method('getSystemValue')
+			->with('version', '0.0.0')
+			->willReturn('10.0.3');
 		$this->repair->run($outputMock);
 
 		// orphaned entry reattached
@@ -701,6 +720,10 @@ class RepairMismatchFileCachePathTest extends TestCase {
 
 		$outputMock = $this->createMock(IOutput::class);
 		$this->repair->setStorageNumericId($storageId);
+		$this->config->expects($this->any())
+			->method('getSystemValue')
+			->with('version', '0.0.0')
+			->willReturn('10.0.3');
 		$this->repair->run($outputMock);
 
 		// orphaned entry with no root reattached
@@ -728,5 +751,31 @@ class RepairMismatchFileCachePathTest extends TestCase {
 		$this->assertEquals((string)$testStorageId, $entry['storage']);
 		$this->assertEquals('', $entry['path']);
 		$this->assertEquals(\md5(''), $entry['path_hash']);
+	}
+
+	public function provideVersions() {
+		return [
+			['10.0.3'],
+			['10.0.4'],
+			['10.0.3.9'],
+		];
+	}
+
+	/**
+	 * @dataProvider provideVersions
+	 * @param $version
+	 */
+	public function testVersions($version) {
+		$this->config->expects($this->any())
+			->method('getSystemValue')
+			->with('version', '0.0.0')
+			->willReturn($version);
+		$outputMock = $this->createMock(IOutput::class);
+		if (\version_compare(\OC::$server->getConfig()->getSystemValue('version', '0.0.0'), $version, '<')) {
+			$outputMock->expects($this->any())
+				->method('info')
+				->with($this->repair->getName() . ' is not executed');
+		}
+		$this->repair->run($outputMock);
 	}
 }

